core/src/main/scala/cats/data/Prod.scala

```diff
@@ -0,0 +1,108 @@
+package cats
+package data
+
+/**
+ * [[Prod]] is a product to two independent functor values.
+ *
+ * See: [[https://www.cs.ox.ac.uk/jeremy.gibbons/publications/iterator.pdf The Essence of the Iterator Pattern]]
+ */
+sealed trait Prod[F[_], G[_], A] {
+  def first: F[A]
+  def second: G[A]
+}
+object Prod extends ProdInstances {
+  def apply[F[_], G[_], A](first0: => F[A], second0: => G[A]): Prod[F, G, A] = new Prod[F, G, A] {
```

Comment:
Just to echo @ceedubs here, would you consider something like this?

```scala
final class Prod[F[_], G[_], A](first0: Eval[F[A]], second0: Eval[F[A]]) {
  def first: F[A] = first0.value
  def second: G[A] = second0.value
}

object Prod extends ProdInstances {
  def apply[F[_], G[_], A](first0: => F[A], second0: => G[A]): Prod[F, G, A] =
    new Prod(Later(first0), Later(second0))
  ...
}
```

The advantage here is that in situations where you already have an `F[A]` or `G[A]` in-hand, you can simply say `new Prod(Now(fa), Now(ga))`. If you want the syntactic niceness you can still say `Prod(..., ...)`.

What do you think?